File: src/main/java/com/embabel/prepper/AdditionalBedrockModels.java
package com.embabel.prepper;


import com.embabel.agent.config.AgentPlatformConfiguration;
import com.embabel.agent.config.models.bedrock.BedrockModels;
import com.embabel.agent.config.models.bedrock.BedrockOptionsConverter;
import com.embabel.agent.config.models.bedrock.EmbabelBedrockProxyChatModelBuilder;
import com.embabel.common.ai.model.Llm;
import com.embabel.common.ai.model.PerTokenPricingModel;
import io.micrometer.observation.ObservationRegistry;
import org.springframework.ai.chat.model.ChatModel;
import org.springframework.ai.chat.observation.ChatModelObservationConvention;
import org.springframework.ai.model.bedrock.autoconfigure.BedrockAwsConnectionProperties;
import org.springframework.ai.model.tool.ToolCallingChatOptions;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.boot.autoconfigure.AutoConfigureBefore;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
import software.amazon.awssdk.regions.providers.AwsRegionProvider;
import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeAsyncClient;
import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeClient;

import java.util.ArrayList;

@Configuration(proxyBeanMethods = false)
@AutoConfigureBefore(AgentPlatformConfiguration.class)
class AdditionalBedrockModels {

    private final AwsCredentialsProvider credentialsProvider;
    private final AwsRegionProvider regionProvider;
    private final BedrockAwsConnectionProperties connectionProperties;
    private final ObjectProvider<ObservationRegistry> observationRegistry;
    private final ObjectProvider<ChatModelObservationConvention> observationConvention;
    private final ObjectProvider<BedrockRuntimeClient> bedrockRuntimeClient;
    private final ObjectProvider<BedrockRuntimeAsyncClient> bedrockRuntimeAsyncClient;

    public AdditionalBedrockModels(AwsCredentialsProvider credentialsProvider, AwsRegionProvider regionProvider, BedrockAwsConnectionProperties bedrockAwsConnectionProperties, ObjectProvider<ObservationRegistry> observationRegistry, ObjectProvider<ChatModelObservationConvention> observationConvention, ObjectProvider<BedrockRuntimeClient> bedrockRuntimeClient, ObjectProvider<BedrockRuntimeAsyncClient> bedrockRuntimeAsyncClient) {
        this.credentialsProvider = credentialsProvider;
        this.regionProvider = regionProvider;
        this.connectionProperties = bedrockAwsConnectionProperties;
        this.observationRegistry = observationRegistry;
        this.observationConvention = observationConvention;
        this.bedrockRuntimeClient = bedrockRuntimeClient;
        this.bedrockRuntimeAsyncClient = bedrockRuntimeAsyncClient;
    }

    ChatModel chatModelOf(String model) {
        var chatModel = new EmbabelBedrockProxyChatModelBuilder()
                .credentialsProvider(credentialsProvider)
                .region(regionProvider.getRegion())
                .timeout(connectionProperties.getTimeout())
                .defaultOptions(ToolCallingChatOptions.builder().model(model).build())
                .observationRegistry(observationRegistry.getIfUnique() != null ? observationRegistry.getIfUnique() : ObservationRegistry.NOOP)
                .bedrockRuntimeClient(bedrockRuntimeClient.getIfAvailable())
                .bedrockRuntimeAsyncClient(bedrockRuntimeAsyncClient.getIfAvailable())
                .build();

        observationConvention.ifAvailable(chatModel::setObservationConvention);

        return chatModel;
    }


    @Bean("bedrockModel-us.amazon.nova-pro-v1:0")
    Llm usNovaPro() {
        var model = "us.amazon.nova-pro-v1:0";

        return new Llm(
                model,
                BedrockModels.PROVIDER,
                chatModelOf(model),
                BedrockOptionsConverter.INSTANCE,
                java.time.LocalDate.parse("2025-03-01"), // todo
                new ArrayList<>(),
                new PerTokenPricingModel(1.0, 1.0) // todo
        );
    }

    @Bean("bedrockModel-us.amazon.nova-lite-v1:0")
    Llm usNovaLite() {
        var model = "us.amazon.nova-lite-v1:0";

        return new Llm(
                model,
                BedrockModels.PROVIDER,
                chatModelOf(model),
                BedrockOptionsConverter.INSTANCE,
                java.time.LocalDate.parse("2025-03-01"), // todo
                new ArrayList<>(),
                new PerTokenPricingModel(1.0, 1.0) // todo
        );
    }
}

File: src/main/kotlin/com/embabel/prepper/AdditionalBedrockModels.kt
package com.embabel.prepper

import com.embabel.agent.config.AgentPlatformConfiguration
import com.embabel.agent.config.models.bedrock.BedrockModels
import com.embabel.agent.config.models.bedrock.BedrockOptionsConverter
import com.embabel.common.ai.model.Llm
import com.embabel.common.ai.model.PerTokenPricingModel
import io.micrometer.observation.ObservationRegistry
import org.springframework.ai.bedrock.converse.BedrockProxyChatModel
import org.springframework.ai.chat.model.ChatModel
import org.springframework.ai.chat.observation.ChatModelObservationConvention
import org.springframework.ai.model.bedrock.autoconfigure.BedrockAwsConnectionProperties
import org.springframework.ai.model.tool.DefaultToolExecutionEligibilityPredicate
import org.springframework.ai.model.tool.ToolCallingChatOptions
import org.springframework.ai.model.tool.ToolCallingManager
import org.springframework.beans.factory.ObjectProvider
import org.springframework.boot.autoconfigure.AutoConfigureBefore
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient
import software.amazon.awssdk.regions.providers.AwsRegionProvider
import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeAsyncClient
import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeClient
import java.time.LocalDate

@Configuration(proxyBeanMethods = false)
@AutoConfigureBefore(AgentPlatformConfiguration::class)
class AdditionalBedrockModels(
    private val credentialsProvider: AwsCredentialsProvider,
    private val regionProvider: AwsRegionProvider,
    private val connectionProperties: BedrockAwsConnectionProperties,
    private val observationRegistry: ObjectProvider<ObservationRegistry>,
    private val observationConvention: ObjectProvider<ChatModelObservationConvention>,
    private val bedrockRuntimeClientProvider: ObjectProvider<BedrockRuntimeClient>,
    private val bedrockRuntimeAsyncClientProvider: ObjectProvider<BedrockRuntimeAsyncClient>
) {

    fun chatModelOf(model: String): ChatModel {
        val bedrockRuntimeClient = bedrockRuntimeClientProvider.getIfAvailable {
            BedrockRuntimeClient.builder()
                .region(regionProvider.region)
                .httpClientBuilder(null)
                .credentialsProvider(credentialsProvider)
                .overrideConfiguration { c -> c.apiCallTimeout(connectionProperties.timeout) }
                .build()
        }

        val bedrockRuntimeAsyncClient = bedrockRuntimeAsyncClientProvider.getIfAvailable {
            BedrockRuntimeAsyncClient.builder()
                .region(regionProvider.region)
                .httpClientBuilder(
                    NettyNioAsyncHttpClient.builder()
                        .tcpKeepAlive(true)
                        .connectionAcquisitionTimeout(java.time.Duration.ofSeconds(30))
                        .maxConcurrency(200)
                )
                .credentialsProvider(this.credentialsProvider)
                .overrideConfiguration { c -> c.apiCallTimeout(connectionProperties.timeout) }
                .build()
        }

        return BedrockProxyChatModel(
            bedrockRuntimeClient,
            bedrockRuntimeAsyncClient,
            ToolCallingChatOptions.builder().model(model).build(),
            observationRegistry.ifUnique ?: ObservationRegistry.NOOP,
            ToolCallingManager.builder().build(),
            DefaultToolExecutionEligibilityPredicate()
        ).apply {
            observationConvention.ifAvailable?.let { setObservationConvention(it) }
        }
    }

    @Bean("bedrockModel-us.amazon.nova-pro-v1:0")
    fun usNovaPro(): Llm {
        val model = "us.amazon.nova-pro-v1:0"
        return Llm(
            model,
            BedrockModels.PROVIDER,
            chatModelOf(model),
            BedrockOptionsConverter,
            LocalDate.parse("2025-03-01"), // todo
            arrayListOf(),
            PerTokenPricingModel(1.0, 1.0) // todo
        )
    }

    @Bean("bedrockModel-us.amazon.nova-lite-v1:0")
    fun usNovaLite(): Llm {
        val model = "us.amazon.nova-lite-v1:0"
        return Llm(
            model,
            BedrockModels.PROVIDER,
            chatModelOf(model),
            BedrockOptionsConverter,
            LocalDate.parse("2025-03-01"), // todo
            arrayListOf(),
            PerTokenPricingModel(1.0, 1.0) // todo
        )
    }
}
